netdb.c

```diff
@@ -0,0 +1,241 @@
+{
+	rtr_gethostbyaddr_t real_gethostbyaddr;
+	struct hostent *hent;
+
+	char ip_addr[INET6_ADDRSTRLEN];
+
+	real_gethostbyaddr = RETRACE_GET_REAL(gethostbyaddr);
+
+	/* get IP address */
+	inet_ntop(type, addr, ip_addr, sizeof(ip_addr));
+
+	hent = real_gethostbyaddr(addr, len, type);
+	if (hent)
+		trace_printf(1, "gethostbyaddr(IP => %s, , type => %s) = [Hostname:%s]\n", ip_addr,
+			type == AF_INET ? "AF_INET" : "AF_INET6",
+			hent->h_name);
+	else
+		trace_printf(1, "gethostbyaddr(IP => %s, , type => %s) = [Error:%s]\n", ip_addr,
+			type == AF_INET ? "AF_INET" : "AF_INET6",
+			hstrerror(h_errno));
+
+	return hent;
+}
+
+RETRACE_REPLACE(gethostbyaddr);
+
+void RETRACE_IMPLEMENTATION(sethostent)(int stayopen)
+{
+	rtr_sethostent_t real_sethostent;
+
+	real_sethostent = RETRACE_GET_REAL(sethostent);
+
+	trace_printf(1, "sethostent(%d)\n", stayopen);
+
+	real_sethostent(stayopen);
+}
+
+RETRACE_REPLACE(sethostent);
+
+void RETRACE_IMPLEMENTATION(endhostent)(void)
+{
+	rtr_endhostent_t real_endhostent;
+
+	real_endhostent = RETRACE_GET_REAL(endhostent);
+
+	trace_printf(1, "endhostent()\n");
+
+	real_endhostent();
+}
+
+RETRACE_REPLACE(endhostent);
+
+struct hostent *RETRACE_IMPLEMENTATION(gethostent)(void)
+{
+	rtr_gethostent_t real_gethostent;
+	struct hostent *hent;
+
+	real_gethostent = RETRACE_GET_REAL(gethostent);
+
+	hent = real_gethostent();
+	if (hent) {
+		struct in_addr **addr_list;
+		int i;
+
+		trace_printf(1, "gethostent() = [");
+
+		addr_list = (struct in_addr **) hent->h_addr_list;
+		for (i = 0; addr_list[i] != NULL; i++) {
+			char ip_addr[INET6_ADDRSTRLEN];
+
+			inet_ntop(hent->h_addrtype, addr_list[i], ip_addr, sizeof(ip_addr));
+			trace_printf(0, i > 0 ? ",%s" : "%s", ip_addr);
+		}
+
+		trace_printf(0, "]\n");
+	} else
+		trace_printf(1, "gethostent() = NULL [error:%s]\n", hstrerror(h_errno));
+
+	return hent;
+}
+
+RETRACE_REPLACE(gethostent);
+
+struct hostent *RETRACE_IMPLEMENTATION(gethostbyname2)(const char *name, int af)
+{
+	rtr_gethostbyname2_t real_gethostbyname2;
+	struct hostent *hent;
+
+	char ip_addr[INET6_ADDRSTRLEN];
+	int i;
+
+	real_gethostbyname2 = RETRACE_GET_REAL(gethostbyname2);
+
+	hent = real_gethostbyname2(name, af);
+	if (hent) {
+		trace_printf(1, "gethostbyname2(\"%s\", %s) = [", name, af == AF_INET ? "AF_INET" : "AF_INET6");
+
+		if (af == AF_INET) {
+			struct in_addr **addr_list;
+
+			addr_list = (struct in_addr **) hent->h_addr_list;
+			for (i = 0; addr_list[i] != NULL; i++) {
+				inet_ntop(hent->h_addrtype, addr_list[i], ip_addr, sizeof(ip_addr));
+				trace_printf(0, i > 0 ? ",%s" : "%s", ip_addr);
+			}
+		} else if (af == AF_INET6) {
+			struct in6_addr **addr_list;
+
+			addr_list = (struct in6_addr **) hent->h_addr_list;
+			for (i = 0; addr_list[i] != NULL; i++) {
+				inet_ntop(hent->h_addrtype, addr_list[i], ip_addr, sizeof(ip_addr));
+				trace_printf(0, i > 0 ? ",%s" : "%s", ip_addr);
+			}
+		}
```

Comment:
@jjr840430 as per previous comment, this if/else is  unnecessary. `addr_list` is an array of pointers, `inet_ntop` does the cast to `in_addr *` or `in6_addr *` depending on `hent->addrtype` so you don't need to:

`inet_ntop(hent->h_addrtype, hent->h_addr_list[i], ip_addr, sizeof(ip_addr));`